In [1]:
import logging
logging.basicConfig(format="%(message)s | %(levelname)s", level=logging.INFO, force=True)

In [2]:
import requests

# working on the get requests

url = "https://jsonplaceholder.typicode.com/todos/1"
raw = requests.get(url)
jsonData = raw.json()
logging.info(jsonData)

{'userId': 1, 'id': 1, 'title': 'delectus aut autem', 'completed': False} | INFO


In [3]:
raw.status_code

200

In [4]:
raw.content

b'{\n  "userId": 1,\n  "id": 1,\n  "title": "delectus aut autem",\n  "completed": false\n}'

In [9]:
list(raw.headers.items())

[('Date', 'Fri, 15 Dec 2023 14:16:01 GMT'),
 ('Content-Type', 'application/json; charset=utf-8'),
 ('Transfer-Encoding', 'chunked'),
 ('Connection', 'keep-alive'),
 ('Report-To',
  '{"group":"heroku-nel","max_age":3600,"endpoints":[{"url":"https://nel.heroku.com/reports?ts=1702559859&sid=e11707d5-02a7-43ef-b45e-2cf4d2036f7d&s=CrkYX1LEyfyCED%2F0JaS4JsVKMpZfuy7FTKQGgNuMg18%3D"}]}'),
 ('Reporting-Endpoints',
  'heroku-nel=https://nel.heroku.com/reports?ts=1702559859&sid=e11707d5-02a7-43ef-b45e-2cf4d2036f7d&s=CrkYX1LEyfyCED%2F0JaS4JsVKMpZfuy7FTKQGgNuMg18%3D'),
 ('Nel',
  '{"report_to":"heroku-nel","max_age":3600,"success_fraction":0.005,"failure_fraction":0.05,"response_headers":["Via"]}'),
 ('X-Powered-By', 'Express'),
 ('X-Ratelimit-Limit', '1000'),
 ('X-Ratelimit-Remaining', '999'),
 ('X-Ratelimit-Reset', '1702559866'),
 ('Vary', 'Origin, Accept-Encoding'),
 ('Access-Control-Allow-Credentials', 'true'),
 ('Cache-Control', 'max-age=43200'),
 ('Pragma', 'no-cache'),
 ('Expires', '-1'),
 (

In [3]:
# work on POST request: where the data is posted

data = {"userId": 2,
        "title": "Buy Milk",
        "completed": False}
post_url = "https://jsonplaceholder.typicode.com/todos"
# resp = requests.post(url=post_url, data=data) # this also works
resp = requests.post(url=post_url, json=data)

In [13]:
resp.status_code

201

In [4]:
resp.json()

{'userId': 2, 'title': 'Buy Milk', 'completed': False, 'id': 201}

In [6]:
# if the json keyword is not used then multiple info has to be provide
headers = {"Content-Type":"application/json"}

# work on posting
new_resp = requests.post(post_url, data=data, headers=headers)
# new_resp.json()
# got the decode error :  Expecting value: line 1 column 1 (char 0)
# reviewing stack overflow https://stackoverflow.com/questions/16573332/jsondecodeerror-expecting-value-line-1-column-1-char-0
# gave me idea that, requests is not getting through
new_resp

<Response [500]>

In [7]:
import json
# if the json keyword is not used then multiple info has to be provide
headers = {"Content-Type":"application/json"}

# work on posting, and the dictionary has to be dumped using json.dumps
new_resp = requests.post(post_url, data=json.dumps(data), headers=headers)
# Now the requests is getting through
new_resp.json()

{'userId': 2, 'title': 'Buy Milk', 'completed': False, 'id': 201}

In [17]:
get10 = "https://jsonplaceholder.typicode.com/todos/10"
resp10 = requests.get(get10)
resp10.json()

{'userId': 1,
 'id': 10,
 'title': 'illo est ratione doloremque quia maiores aut',
 'completed': True}

In [9]:
# modifying the data using put
updated_data = {'userId': 1,
 'id': 10,
 'title': 'Make arrangements for the delivery',
 'completed': False}

In [12]:
resp_put = requests.put(url=get10, json=updated_data)
resp_put.json()

{'userId': 1,
 'id': 10,
 'title': 'Make arrangements for the delivery',
 'completed': False}

In [13]:
part_data = {"title":"move the vehicle"}
patch_resp = requests.patch(get10, json=part_data)
patch_resp.json()
# in case of the json placeholder, the modifications are not getting 
# registered in the backend

{'userId': 1, 'id': 10, 'title': 'move the vehicle', 'completed': True}

In [16]:
# deleting
resp_delete = requests.delete(url=get10)
resp_delete.json()

{}

### Process for creating defining APIs

- decide the resources that API will be managing

- list the endpoints and http methods, get, post, put, patch, and delete

HTTP method	    API endpoint	                        Description

GET	            /events/<event_id>/guests	            Get a list of guests.

GET	            /events/<event_id>/guests/<guest_id>	Get a single guest.

POST	        /events/<event_id>/guests	            Create a new guest.

PUT	            /events/<event_id>/guests/<guest_id>	Update a guest.

PATCH	        /events/<event_id>/guests/<guest_id>	Partially update a guest.

DELETE	        /events/<event_id>/guests/<guest_id>	Delete a guest.

# HTTP requests format

GET /cars HTTP/1.1

Host: api.example.com

# GET response format

HTTP/1.1 200 OK
Content-Type: application/json

[
    {"Json": "data"}
]

# HTTP Post request format

POST /cars HTTP/1.1

Host: api.example.com

Content-Type: application/json

{"make": "Nissan",

    "model": "240SX",
    
    "year": 1994,
    
    "vin": "1N6AD0CU5AC961553",
    
    "color": "Violet"}

### Put request also is similar, only difference is the data is updated completely

### Patch request also is similar, in which a part of data is updated.

### All three methods will send the copy of the data back

# HTTP Post Response format

Note, a copy of the data is sent back in the response

HTTP/1.1 201 Created
Content-Type: application/json

{"make": "Nissan",

    "model": "240SX",
    
    "year": 1994,
    
    "vin": "1N6AD0CU5AC961553",
    
    "color": "Violet"}

# HTTP Delete request 

DELETE /cars/4 HTTP/1.1

# HTTP Delete response

HTTP/1.1 204 No Content

# Some Errors

In [18]:
no_url = "https://jsonplaceholder.typicode.com/cars/1"

# trying to get a resource that is not available.

resp = requests.get(url=no_url)
resp.json()

{}

In [19]:
resp.status_code

404

In [28]:
list(resp.headers.items())

[('Date', 'Sat, 16 Dec 2023 04:27:55 GMT'),
 ('Content-Type', 'application/json; charset=utf-8'),
 ('Content-Length', '2'),
 ('Connection', 'keep-alive'),
 ('Report-To',
  '{"group":"heroku-nel","max_age":3600,"endpoints":[{"url":"https://nel.heroku.com/reports?ts=1702700875&sid=e11707d5-02a7-43ef-b45e-2cf4d2036f7d&s=ObUSqGwhSWt6%2BU9jo7zGUJ1wmNys0Hj6xqsriBFOFpM%3D"}]}'),
 ('Reporting-Endpoints',
  'heroku-nel=https://nel.heroku.com/reports?ts=1702700875&sid=e11707d5-02a7-43ef-b45e-2cf4d2036f7d&s=ObUSqGwhSWt6%2BU9jo7zGUJ1wmNys0Hj6xqsriBFOFpM%3D'),
 ('Nel',
  '{"report_to":"heroku-nel","max_age":3600,"success_fraction":0.005,"failure_fraction":0.05,"response_headers":["Via"]}'),
 ('X-Powered-By', 'Express'),
 ('X-Ratelimit-Limit', '1000'),
 ('X-Ratelimit-Remaining', '999'),
 ('X-Ratelimit-Reset', '1702700927'),
 ('Vary', 'Origin, Accept-Encoding'),
 ('Access-Control-Allow-Credentials', 'true'),
 ('Cache-Control', 'max-age=43200'),
 ('Pragma', 'no-cache'),
 ('Expires', '-1'),
 ('X-Conten

In [26]:
# trying to post wrongly

# data_mal = {'id':2, 'title':'lets break stuff'
            
mal_resp = requests.post(url=no_url, json={'id':2, 'title':'lets break stuff'})
mal_resp.status_code

404

#### trying the get request in shell using curl will be 

curl -i https://jsonplaceholder.typicode.com/todos -X GET


curl -i https://jsonplaceholder.typicode.com/todos -X POST \
-H 'Content-Type: application/json' \
-d '{"id":10, "title":"using curl", "completed":False} 